q2_sample_classifier/plugin_setup.py

```diff
@@ -57,6 +59,102 @@ def sniff(self):
     BooleanSeriesFormat)
 
 
+Predictions = SemanticType(
+    'Predictions', variant_of=SampleData.field['type'])
+
+
+def _validate_file_not_empty(has_data):
+    if not has_data:
+        raise ValidationError(
+            "There must be at least one data record present in the "
+            "file in addition to the header line.")
+
+
+class PredictionsFormat(model.TextFileFormat):
+    def _validate(self, n_records=None):
+        with self.open() as fh:
+            # validate header
+            # for now we will not validate any information in the header,
+            # since the name of the predicted column should be flexible. The
+            # header name written by methods in q2-sample-classifier will be
+            # "predicted-*", but this should also accommodate user-defined
+            # column names.
+            line = fh.readline()
+
+            # validate body
+            has_data = False
+            for line_number, line in enumerate(fh, start=2):
+                cells = line.strip().split('\t')
+                if len(cells) != 2:
+                    raise ValidationError(
+                        "Expected data record to be TSV with two "
+                        "fields. Detected {0} fields at line {1}:\n\n{2!r}"
+                        .format(len(cells), line_number, cells))
+                has_data = True
+                if n_records is not None and (line_number - 1) >= n_records:
+                    break
+
+            _validate_file_not_empty(has_data)
+            return True
```

Comment:
The semantics of sniffing versus validation here are a bit different - validation will either complete successfully, or will raise an exception. This is a "return-less" operation (or, `return None`, which python will do without an explicit return). Anyway, for the sake of future readers, it might be misleading to see this, since nowhere in the validation stack does the return value get checked.

Reply:
👍